src/Verify.MSTest.SourceGenerator/Parser.cs

```diff
@@ -0,0 +1,102 @@
+namespace VerifyMSTest.SourceGenerator;
+
+static class Parser
+{
+    public static string MarkerAttributeName => "VerifyMSTest.UsesVerifyAttribute";
+
+    public static ClassToGenerate? Parse(INamedTypeSymbol typeSymbol, TypeDeclarationSyntax typeSyntax)
+    {
+        // Only generate for classes that don't already have a TestContext property defined.
+        if (HasTestContextProperty(typeSymbol))
+        {
+            return null;
+        }
+
+        var ns = GetNamespace(typeSymbol);
+        var name = GetTypeNameWithGenericParameters(typeSyntax);
+        var parents = GetParentClasses(typeSyntax);
+
+        return new ClassToGenerate(ns, name, parents);
+    }
+
+    static string? GetNamespace(INamedTypeSymbol symbol) =>
+        symbol.ContainingNamespace.IsGlobalNamespace ? null : symbol.ContainingNamespace.ToString();
+
+    static bool HasTestContextProperty(INamedTypeSymbol symbol) =>
+        HasMarkerAttributeOnBase(symbol) ||
+        HasTestContextPropertyDefinedInBase(symbol);
+
+    static bool HasMarkerAttributeOnBase(INamedTypeSymbol symbol)
+    {
+        static bool HasMarkerAttribute(ISymbol symbol) =>
+            symbol
+            .GetAttributes()
+            .Any(a => a.AttributeClass?.ToDisplayString() == MarkerAttributeName);
+
+        var parent = symbol.BaseType;
+
+        while (parent is not null)
+        {
+            if (HasMarkerAttribute(parent))
+            {
+                return true;
+            }
+
+            parent = parent.BaseType;
+        }
+
+        return false;
+    }
+
+    static bool HasTestContextPropertyDefinedInBase(INamedTypeSymbol symbol)
+    {
+        static bool HasTestContextProperty(INamedTypeSymbol symbol) =>
+            symbol
+            .GetMembers()
+            .OfType<IPropertySymbol>()
+            .Any(_ => _.Name == "TestContext" &&
+                      _.Type.Name == "TestContext");
```

Comment:
should we check namespace?

Reply:
Now that you mention it, the type doesn't matter at all since you can't overload a property based on the return type. Removed the type check.